---
title: "netOmics"
author: 
- name:  "Antoine Bodein"
  affiliation: "CHU de Québec Research Center, Université Laval, Molecular Medicine department, Québec, QC, Canada"
  email: "antoine.bodein.1@ulaval.ca"
- name:  "Marie-Pier Scott-Boyer"
  affiliation: "CHU de Québec Research Center, Université Laval, Molecular Medicine department, Québec, QC, Canada"
- name: "Olivier Perin"
  affiliation: "Digital Sciences Department, L’Oréal Advanced Research, Aulnay-sous-bois, France"
- name: "Kim-Anh Lê Cao"
  affiliation: "Melbourne Integrative Genomics, School of Mathematics and Statistics, University of Melbourne, Melbourne, VIC, Australia"
- name: "Arnaud Droit"
  affiliation: "CHU de Québec Research Center, Université Laval, Molecular Medicine department, Québec, QC, Canada"
package: netOmics  
output: 
  BiocStyle::html_document
  
vignette: >
  %\VignetteIndexEntry{timeOmics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  

---

In this example, we will describe step by step how to create multi-omics networks and sub-networks by kinetic cluster from multi-omics time course data.

```{r, eval=TRUE}
library(netOmics)
library(timeOmics)
library(tidyverse)

```

# Case Study: Human Microbiome Project T2D

```{r load_data}
data("hmp_T2D")
```


# timeOmics 
```{r}
# clustering

cluster.info <- hmp_T2D$getCluster.res
```

# network building

## Gene regulatory network

We start building the multi-omics network with a first layer of gene.
Currently, the ARACNe algorithm handles the inference but we will include more algorithms in the future.

The function `get_grn` return a Gene Regulatory Network from gene expression data. 
Optionally, the user can provide a timeOmics clustering result to get cluster specific sub-networks.

The `get_graph_stats` function provides basic graph statistics such as the number of verticesIl inclu egale,m and nodes.
If the vertices have different attributes, it also includes a summary of these.
```{r rna.graph}
cluster.info.RNA <- timeOmics::getCluster(cluster.info, user.block = "RNA")
rna.graph <- get_grn(X = hmp_T2D$data$RNA, cluster = cluster.info.RNA)
get_graph_stats(rna.graph)
```

## PPI

As for the genes, we will make a protein layer (Protein-Protein Interaction).
This time we will extract the interaction from a database (BIOGRID).

The function `get_interaction_from_database` will fetch the interactions from a database provided as a `data.frame` (with columns `from` and `to`) or a graph (`igraph` object).
In addition to the interactions between the indicated molecules, the first degree neighbours can also be collected (option `user.ego = TRUE`)

```{r}
# Utility function to get the molecules by cluster
get_list_mol_cluster <- function(cluster.info, user.block){
  require(timeOmics)
    tmp <- timeOmics::getCluster(cluster.info, user.block) 
    res <- tmp %>% split(.$cluster) %>% 
        lapply(function(x) x$molecule)
    res[["All"]] <- tmp$molecule
    return(res)
}

cluster.info.prot <- get_list_mol_cluster(cluster.info, user.block = 'PROT')
prot.graph <-  get_interaction_from_database(X = cluster.info.prot, db = hmp_T2D$interaction.biogrid, type = "PROT", user.ego = TRUE)
get_graph_stats(prot.graph)
```
## GUT
```{r, eval = FALSE}
library(SpiecEasi)

get_sparcc_graph <- function(X, threshold = 0.3){
    res.sparcc <- sparcc(data = X)
    sparcc.graph <- abs(res.sparcc$Cor) >= threshold
    colnames(sparcc.graph) <-  colnames(X)
    rownames(sparcc.graph) <-  colnames(X)
    res.graph <- graph_from_adjacency_matrix(sparcc.graph, mode = "undirected") %>% simplify
    return(res.graph)
}

cluster.info.gut <- timeOmics::getCluster(X = cluster.info, user.block = 'GUT')

graph.gut <- list()
graph.gut[["All"]] <- get_sparcc_graph(GUT.RAW, threshold = 0.3)
graph.gut[["1"]] <- get_sparcc_graph(GUT.RAW %>% dplyr::select(gut_list[["1"]]), threshold = 0.3)
graph.gut[["-1"]] <- get_sparcc_graph(GUT.RAW %>% dplyr::select(gut_list[["-1"]]), threshold = 0.3)
class(graph.gut) <- "list.igraph"

get_graph_stats(graph.gut)

```


## CYTO
```{r}
cluster.info.cyto <- timeOmics::getCluster(X = cluster.info, user.block = 'CYTO')
cyto.graph <-  get_interaction_from_database(X = cluster.info.cyto$molecule, db = hmp_T2D$interaction.biogrid, type = "CYTO", user.ego = FALSE)
get_graph_stats(prot.graph)
```



```{r, eval = FALSE, echo = FALSE}
library(tidyverse)
library(igraph)
load("/home/antoine/Documents/timeomics_analysis/HMP_seasoning/netomics/timeomics_res_block.rda")
load("/home/antoine/Documents/timeomics_analysis/HMP_seasoning/netomics/LMMS.RDA")
load("/home/antoine/Documents/timeomics_analysis/HMP_seasoning/netomics/DATA_FILTERED.RDA")
load("/home/antoine/Documents/timeomics_analysis/HMP_seasoning/netomics/RAW_DATA.RDA")

biogrid <- read_tsv("/home/antoine/Documents/TO2/netOmics-case-studies/HeLa_Cell_Cycling/data/BIOGRID-ALL-3.5.187.tab3.txt")
biogrid.filtered <- biogrid %>% dplyr::select("Official Symbol Interactor A", "Official Symbol Interactor B") %>% unique %>%
    set_names(c("from", "to"))

TFome <- readRDS( "~/Documents/TO2/TFome.Rds")

get_list_mol_cluster <- function(cluster.info, user.block){
    #timeOmics::getCluster(cluster.info)
    tmp <- cluster.info %>% filter(block %in% user.block)
    res <- tmp %>% split(.$cluster) %>% 
        lapply(function(x) x$molecule)
    res[["All"]] <- tmp$molecule
    return(res)
}

cluster.info <- timeOmics::getCluster(X = block.res)

### Genes
# Build network from raw data
cluster.info.mRNA <- timeOmics::getCluster(X = cluster.info, user.block = 'RNA')
rna.graph <- get_grn(DATA.GOOD$RNA, cluster = cluster.info.mRNA)
get_graph_stats(rna.graph)


### prot graph
# names are SYMBOL
cluster.info.prot <- timeOmics::getCluster(X = cluster.info, user.block = 'PROT')
prot_list = get_list_mol_cluster(cluster.info = cluster.info, user.block = "PROT")
prot.graph <-  get_interaction_from_database(X = prot_list, db = BIOGRID, type = "PROT", user.ego = TRUE)

### CYTO # same as biogrid (cytokines are proteins)
# names are SYMBOL
cluster.info.cyto <- timeOmics::getCluster(X = cluster.info, user.block = 'CYTO')
cyto_list = get_list_mol_cluster(cluster.info = cluster.info, user.block = "CYTO")
cyto.graph <-  get_interaction_from_database(X = cyto_list, db = biogrid.filtered, type = "CYTO", user.ego = FALSE)
get_graph_stats(cyto.graph)
#lapply(biogrid, function(x) any(cyto_list$All %in% x))
cyto.graph <-  get_interaction_from_database(X = cyto_list, db = biogrid.filtered, type = "CYTO", user.ego = TRUE)
get_graph_stats(cyto.graph)

# all_mol_ego <- unique(c(V(cyto.graph$All)$name, V(rna.graph$All)$name, V(prot.graph$All)$name))

# TFome
# TFome <- readRDS( "~/Documents/TO2/TFome.Rds")
# TFome <- filter(TFome, TFome$TF %in% all_mol_ego | TFome$Target %in% all_mol_ego) %>% 
    # unique
# coding
# AnnotationDbi::select(org.Hs.eg.db, keytype = "SYMBOL", keys = all_mol_ego,
#                       columns = c("SYMBOL", "UNIPROT")) %>%
#     unique %>% na.omit
# 
# AnnotationDbi::select(org.Hs.eg.db, keytype = "SYMBOL", keys = prot_list$All,
#                       columns = c("SYMBOL", "UNIPROT")) %>%
#     unique %>% na.omit


### GUT
library(SpiecEasi)
GUT.RAW <- DATA$GUT.IS[rownames(DATA$GUT.IS) %in% rownames(DATA.GOOD$GUT), ] %>% 
    dplyr::select(names(DATA.GOOD$GUT))
# GUT.RAW <- GUT.RAW[,colSums(GUT.RAW) != 0]

get_sparcc_graph <- function(X, threshold = 0.3){
    res.sparcc <- sparcc(data = X)
    sparcc.graph <- abs(res.sparcc$Cor) >= threshold
    colnames(sparcc.graph) <-  colnames(X)
    rownames(sparcc.graph) <-  colnames(X)
    res.graph <- graph_from_adjacency_matrix(sparcc.graph, mode = "undirected") %>% simplify
    return(res.graph)
}

combine_layers(graph1 = rna.graph, graph2 = prot.graph)

# tmp <- graph_from_adjacency_matrix(sparcc.graph, mode = "undirected") %>% simplify
# plot(tmp, vertex.label = NA, vertex.size = 5)
# degree(tmp) %>% hist

gut_list <- get_list_mol_cluster(cluster.info, "GUT")

graph.gut <- list()
graph.gut[["All"]] <- get_sparcc_graph(GUT.RAW, threshold = 0.3)
graph.gut[["1"]] <- get_sparcc_graph(GUT.RAW %>% dplyr::select(gut_list[["1"]]), threshold = 0.3)
graph.gut[["-1"]] <- get_sparcc_graph(GUT.RAW %>% dplyr::select(gut_list[["-1"]]), threshold = 0.3)
class(graph.gut) <- "list.igraph"

get_graph_stats(graph.gut)


### Clinical
clinical_list <- get_list_mol_cluster(cluster.info, "CLINICAL")
cluster.info.clinical <- timeOmics::getCluster(X = cluster.info, user.block = 'CLINICAL')
graph.clinical <- get_grn(X = DATA.GOOD$CLINICAL, cluster = cluster.info.clinical)
get_graph_stats(graph.clinical)

get_corr_graph <- function(X, threshold = 0.85){
    res.corr <- cor(x = X, method = "spearman")
    corr.graph <- abs(res.corr) >= threshold
    res.graph <- graph_from_adjacency_matrix(corr.graph, mode = "undirected") %>% simplify
    return(res.graph)
}

# graph.clinical <- list()
# graph.clinical[["All"]] <- get_corr_graph(DATA.GOOD$CLINICAL, threshold = 0.2)
# graph.clinical[["1"]] <-   get_corr_graph(DATA.GOOD$CLINICAL%>% dplyr::select(clinical_list[["1"]]), threshold = 0.2)
# graph.clinical[["-1"]] <-  get_corr_graph(DATA.GOOD$CLINICAL%>% dplyr::select(clinical_list[["-1"]]), threshold = 0.2)
# get_graph_stats(graph.clinical)

### Metab
cluster.info.metab <- timeOmics::getCluster(X = cluster.info, user.block = 'METAB')
graph.metab <- get_grn(X = DATA.GOOD$METAB, cluster = cluster.info.metab)
get_graph_stats(graph.metab)


# Merging -----------------------------------------------------------------
## rna, prot, cyto with tfome
full.graph <- combine_layers(graph1 = rna.graph, graph2 = prot.graph)
full.graph <- combine_layers(graph1 = full.graph, graph2 = cyto.graph)

TFome.igraph <- graph_from_data_frame(TFome, directed = FALSE)
TF.interact <- get_interaction_from_database(X = purrr::map(full.graph, ~V(.x)$name), db = TFome.igraph, user.ego = FALSE, type = "TF")

full.graph <- combine_layers(graph1 = full.graph, graph2 = TF.interact)
get_graph_stats(full.graph)

## connecting layers based on correlation
# gut.corr.interact <- get_correlation_incidence(X = DATA.GOOD$GUT, Y = DATA.GOOD, threshold = 0.95)
# clinical.corr.interact <- get_correlation_incidence(X = DATA.GOOD$CLINICAL, Y = DATA.GOOD, threshold = 0.95)
# metab.corr.interact <- get_correlation_incidence(X = DATA.GOOD$METAB, Y = DATA.GOOD, threshold = 0.95)

# matching rownames / samples names
# merged_data <- do.call(DATA.GOOD, what = "cbind")  # pb with colnames -> RNA. ...
# com <- function(x,y) sum(x %in% y)
# com_list <- function(data_list1, data_list2 = data_list1) sapply(data_list1, function(x) sapply(data_list2, function(y) com(x, y)))
# com_list(lapply(DATA.GOOD, names))

merged_data <- reduce(DATA.GOOD, cbind)
rna_prot_cyto_list <- get_list_mol_cluster(cluster.info, user.block = c("RNA", "PROT", "CYTO"))
merged_data_1 <- lapply(rna_prot_cyto_list, function(x)dplyr::select(merged_data, x))

gut_list <- get_list_mol_cluster(cluster.info, user.block = "GUT")
gut_data_cluster <- lapply(gut_list, function(x)dplyr::select(DATA.GOOD$GUT, x))

interaction_df_gut <- get_correlation_incidence_v2(X = gut_data_cluster, Y = merged_data_1, threshold = 0.99)
full.graph <- combine_layers(graph1 = full.graph, graph2 = graph.gut, interaction.df = interaction_df_gut$All)


#### Clinical
rna_prot_cyto_list_gut <- get_list_mol_cluster(cluster.info, user.block = c("RNA", "PROT", "CYTO", "GUT"))
merged_data_1 <- lapply(rna_prot_cyto_list_gut, function(x)dplyr::select(merged_data, x))

clinical_list <- get_list_mol_cluster(cluster.info, user.block = "CLINICAL")
clinical_data_cluster <- lapply(clinical_list, function(x)dplyr::select(DATA.GOOD$CLINICAL, x))

interaction_df_clinical <- get_correlation_incidence_v2(X = clinical_data_cluster, Y = merged_data_1, threshold = 0.99)
full.graph <- combine_layers(graph1 = full.graph, graph2 = graph.clinical, interaction.df = interaction_df_clinical$All)


### Metab
rna_prot_cyto_list_gut_clinical <- get_list_mol_cluster(cluster.info, user.block = c("RNA", "PROT", "CYTO", "GUT", "CLINICAL"))
merged_data_1 <- lapply(rna_prot_cyto_list_gut_clinical, function(x)dplyr::select(merged_data, x))

metab_list <- get_list_mol_cluster(cluster.info, user.block = "METAB")
metab_data_cluster <- lapply(metab_list, function(x)dplyr::select(DATA.GOOD$METAB, x))

interaction_df_metab <- get_correlation_incidence_v2(X = metab_data_cluster, Y = merged_data_1, threshold = 0.99)
full.graph <- combine_layers(graph1 = full.graph, graph2 = graph.metab, interaction.df = interaction_df_metab$All)

# ora analysis
get_ora <- function(X){
    #tmp <- combined_res$BV_10e3_PC2.nolines.spca.comp1 %>% split(.$cluster)
    tmp <- X %>% split(.$cluster)
    tmp[["All"]] <- X
    
    tmp.res <- list()
    for(tmp.cluster in names(tmp)){
        x.query <- tmp[[tmp.cluster]]$molecule
        gp.res <- gprofiler2::gost(query = x.query, organism = "hsapiens", significant = TRUE, sources = c("GO"))
        res <- gp.res$result
        if(!is.null(res)){
            res <- res %>% mutate(cluster = paste0("cluster ", tmp.cluster)) %>% 
                dplyr::select("cluster", "term_id","source","term_name","p_value","significant", "term_size", "query_size", "intersection_size", "precision", "recall")
        } 
        tmp.res[[tmp.cluster]] <- res
    }
    RES <- map_dfr(tmp.res, ~.x)
    
    return(RES)
}

get_ora_by_cluster_and_omics <- function(X){
    #tmp <- combined_res$BV_10e3_PC2.nolines.spca.comp1 %>% split(.$cluster)
    tmp <- X %>% split(.$cluster)
    tmp[["All"]] <- X
    
    tmp.res <- list()
    for(tmp.cluster in names(tmp)){
        # print(tmp.cluster)
        tmp.res.2 <- list()
        for(omic in c("RNA", "PROT", "CYTO")){
            x.query <- tmp[[tmp.cluster]] %>% filter(block == omic) %>% pull(molecule)
            gp.res <- gprofiler2::gost(query = x.query, organism = "hsapiens", significant = TRUE, sources = c("GO"))
            res <- gp.res$result
            if(!is.null(res)){
                res <- res %>% mutate(cluster = paste0("cluster ", tmp.cluster)) %>% 
                    mutate(OMIC = omic) 
                # print(omic)
                tmp.res.2[[omic]] <- res
            }
        }
        tmp.res[[tmp.cluster]] <- map_dfr(tmp.res.2, ~.x)
    }
    RES <- map_dfr(tmp.res, ~.x)
    
    return(RES)
}
    
# x.query <- cluster.info %>% filter(block == "PROT", cluster == 1) %>% pull(molecule)
# gp.res <- gprofiler2::gost(query = x.query, organism = "hsapiens", significant = TRUE, sources = c("GO"))

ORA.res <- get_ora(cluster.info)
ORA.res_by_cluster_omic <- get_ora_by_cluster_and_omics(cluster.info)
goterm_map <- gprofiler2::gconvert(query = cluster.info$molecule, organism = "hsapiens", target = "GO") 

com <- function(x,y) sum(x %in% y)
not_com <- function(x,y) sum(!(x %in% y))
com_list <- function(data_list1, data_list2 = data_list1) sapply(data_list1, function(x) sapply(data_list2, function(y) com(x, y)))
not_com_list <- function(data_list1, data_list2 = data_list1) sapply(data_list1, function(x) sapply(data_list2, function(y) not_com(x, y)))


unique_GO <- ORA.res_by_cluster_omic %>% dplyr::select(cluster, term_id) %>% unique %>% split(.$cluster) %>% map(~.x %>% pull(term_id))
com_list(unique_GO)
not_com_list(unique_GO)


signif_goterm_map <- goterm_map %>% filter(target %in% (ORA.res %>% filter(significant == TRUE) %>% pull(term_id))) %>%
    dplyr::select(input, target) %>% unique %>% na.omit()

graph.go <- list()
graph.go[['All']] <- igraph::graph_from_data_frame(signif_goterm_map, directed = FALSE)
graph.go[['1']] <-  igraph::graph_from_data_frame(signif_goterm_map %>% filter(input %in% (cluster.info %>% filter(cluster == 1) %>% pull(molecule))), directed = FALSE)
graph.go[['-1']] <-  igraph::graph_from_data_frame(signif_goterm_map %>% filter(input %in% (cluster.info %>% filter(cluster == -1) %>% pull(molecule))), directed = FALSE)
class(graph.go) <- "list.igraph"

full.graph <- combine_layers(graph1 = full.graph, graph2 = graph.go)

# # au passage, mol without go
any(!(rna_prot_cyto_list$All %in% unique(goterm_map$input)))

# medline ranker
## commented to not overwrite
# cluster.info %>% filter(block %in% c("RNA", "PROT", "CYTO")) %>% select(molecule) %>%  
#     write_csv("/home/antoine/Documents/timeomics_analysis/HMP_seasoning/netomics/mol_for_medline_ranker_all.csv", quote = FALSE, col_names = FALSE)
## medline ranker
# http://cbdm-01.zdv.uni-mainz.de/~jfontain/cms/?page_id=592

medlineranker.res <- read_tsv("/home/antoine/Documents/timeomics_analysis/HMP_seasoning/netomics/res_medlineranker_all.csv")
tmp <- dplyr::select(medlineranker.res, c("Disease", 'Gene symbols'))  %>% set_names(c("Disease", "symbol")) 
# separate_rows(tmp, Disease, symbol, sep = "|", convert = TRUE)
    
library(splitstackshape)
tmp[c(1,2),] %>% separate_rows(Disease, symbol)
tmp[c(2,3),] %>% 
    rownames_to_column("id") %>% 
    cSplit(., sep = "|", splitCols = 2:ncol(.), direction = "long", makeEqual = T) %>% 
    as_tibble() %>% 
    group_by(id) %>% 
    fill(2:ncol(.)) %>% 
    unique() %>% 
    ungroup(id) %>% 
    select(-id)

medlineranker.res.df <- tmp %>% rownames_to_column("id") %>% 
    cSplit(., sep = "|", splitCols = 3:ncol(.), direction = "long", makeEqual = T) %>% 
    as_tibble() %>% group_by(id) %>% fill(2:ncol(.)) %>% unique() %>% ungroup(id) %>% select(-id)

# full.graph <- combine_layers(graph1 = full.graph, interaction.df = medlineranker.res.df)

graph.medlineranker <- list()
graph.medlineranker[['All']] <- igraph::graph_from_data_frame(medlineranker.res.df, directed = FALSE)
graph.medlineranker[['1']] <-  igraph::graph_from_data_frame(medlineranker.res.df %>% filter(symbol %in% (cluster.info %>% filter(cluster == 1) %>% pull(molecule))), directed = FALSE)
graph.medlineranker[['-1']] <-  igraph::graph_from_data_frame(medlineranker.res.df %>% filter(symbol %in% (cluster.info %>% filter(cluster == -1) %>% pull(molecule))), directed = FALSE)
class(graph.medlineranker) <- "list.igraph"

full.graph <- combine_layers(graph1 = full.graph, graph2 = graph.medlineranker)


# graph cleaning
get_graph_stats(full.graph)
graph_cleaning <- function(X, cluster.info){
    # no reusability
    X <- igraph::simplify(X)
    va <- vertex_attr(X)
    viewed_mol <- c()
    for(omic in unique(cluster.info$block)){
        mol <- intersect(cluster.info %>% filter(.$block == omic) %>% pull(molecule), V(X)$name)
        viewed_mol <- c(viewed_mol, mol)
        X <- set_vertex_attr(graph = X, name = "type", index = mol, value = omic)
        X <- set_vertex_attr(graph = X, name = "mode", index = mol, value = "core")
    }
    # add medline ranker and go
    mol <- intersect(goterm_map$target, V(X)$name)
    viewed_mol <- c(viewed_mol, mol)
    X <- set_vertex_attr(graph = X, name = "type", index = mol, value = "GO")
    X <- set_vertex_attr(graph = X, name = "mode", index = mol, value = "extended")
    
    mol <- intersect(as.character(medlineranker.res.df$Disease), V(X)$name)
    viewed_mol <- c(viewed_mol, mol)
    X <- set_vertex_attr(graph = X, name = "type", index = mol, value = "Disease")
    X <- set_vertex_attr(graph = X, name = "mode", index = mol, value = "extended")
    
    other_mol <- setdiff(V(X), viewed_mol)
    if(!is_empty(other_mol)){
        X <- set_vertex_attr(graph = X, name = "mode", index = other_mol, value = "extended")
    }
    X <- set_vertex_attr(graph = X, name = "mode", index = intersect(cluster.info$molecule, V(X)$name), value = "core")
    return(X)
}

# saveRDS(full.graph, file = "/home/antoine/Documents/timeomics_analysis/HMP_seasoning/netomics/full-graph.rds")
FULL <- lapply(full.graph, function(x) graph_cleaning(x, cluster.info))

seeds <- V(FULL$All)$name
rwr_res <- random_walk_restart(FULL, seeds)


# saveRDS(rwr_res, file = "/home/antoine/Documents/timeomics_analysis/HMP_seasoning/netomics/RWR-res.rds")
rwr_res <-  readRDS("/home/antoine/Documents/timeomics_analysis/HMP_seasoning/netomics/RWR-res.rds")

vertex_attr(FULL$All) %>% names
vertex_attr(FULL$All) %>% as.data.frame() %>% pull(type) %>% table 



rwr_type_k15 <- rwr_find_seeds_between_attributes(X = rwr_res, attribute = "type")
rwr_cluster_k15 <- rwr_find_seeds_between_attributes(X = rwr_res, attribute = "cluster")

# save(rwr_res, rwr_type_k15, rwr_cluster_k15, file = "/home/antoine/Documents/timeomics_analysis/HMP_seasoning/netomics/RWR-res.rda")
load("/home/antoine/Documents/timeomics_analysis/HMP_seasoning/netomics/RWR-res.rda")

rwr_type_k15$All$A2M.AS1 %>% vertex_attr() %>% as.data.frame() %>% mutate(seed = .y)
stat_rwr_1_tmp <- imap_dfr(rwr_type_k15$All, ~ vertex_attr(.x) %>% as.data.frame() %>%mutate(seed = .y)) %>%
    group_by(seed, type) %>% summarise(N=n()) %>% spread(type, N, fill = 0)

stat_rwr_2_tmp <- stat_rwr_1_tmp %>% gather(omics, N, -seed) %>% filter(N >0) %>% group_by(seed) %>% summarise(N = n()) %>% 
    arrange(desc(N)) %>%  left_join(vertex_attr(FULL$All) %>% as.data.frame(), by = c("seed"='name'))
# number of distinc omics


stat_rwr_2_tmp %>% filter(type == "Disease")
# add GO Terms Label
get_go_info(stat_rwr_2_tmp$seed)


stat_rwr_2_tmp %>% left_join(get_go_info(stat_rwr_2_tmp$seed), by = c("seed"='GOID')) %>% filter(type == "GO") %>% View
stat_rwr_2_tmp %>% left_join(get_go_info(stat_rwr_2_tmp$seed), by = c("seed"='GOID')) %>% filter(type == "GO", ONTOLOGY ==  "BP") %>% View
stat_rwr_2_tmp %>% left_join(get_go_info(stat_rwr_2_tmp$seed), by = c("seed"='GOID')) %>% filter(type == "GO", ONTOLOGY ==  "MF") %>% View

stat_rwr_2_tmp %>% left_join(get_go_info(stat_rwr_2_tmp$seed), by = c("seed"='GOID')) %>% filter(type == "GO", ONTOLOGY ==  "CC") %>% View

stat_rwr_1_tmp %>%  filter(seed %in% (stat_rwr_2_tmp %>% filter(N == 5) %>% pull(seed)))

stat_rwr_1_tmp %>% filter(GUT > 0) %>% View()

go_o_i <- c("GO:0000902", "GO:0050839", "GO:0005198", "GO:0030020", "GO:0005911", "GO:0005925")
stat_rwr_1_tmp %>%  filter(seed %in% go_o_i) %>% left_join(get_go_info(stat_rwr_2_tmp$seed), by = c("seed"='GOID')) %>% View()


imap_dfr(rwr_cluster_k15$All, ~ vertex_attr(.x) %>% as.data.frame() %>%mutate(seed = .y)) %>% 
    #group_by(seed, cluster) %>% summarise(N = n()) %>% spread(cluster, N)
    group_by(seed, cluster)

```





